core/src/main/java/it/pagopa/selfcare/mscore/core/strategy/CreateInstitutionStrategyIpa.java

```diff
@@ -44,42 +44,49 @@ public CreateInstitutionStrategyIpa(PartyRegistryProxyConnector partyRegistryPro
     @Override
     public Institution createInstitution(CreateInstitutionStrategyInput strategyInput) {
 
-        final InstitutionPaSubunitType subunitType = strategyInput.getSubunitType();
-
-        InstitutionProxyInfo institutionProxyInfo = partyRegistryProxyConnector.getInstitutionById(strategyInput.getTaxCode());
-        CategoryProxyInfo categoryProxyInfo = partyRegistryProxyConnector.getCategory(institutionProxyInfo.getOrigin(), institutionProxyInfo.getCategory());
-
-        Institution institutionToReturn;
-        Optional<Institution> opt = institutionConnector.findByExternalId(strategyInput.getTaxCode());
-        if(opt.isEmpty()){
-            try {
-                Institution institutionEC = getInstitutionEC(strategyInput.getTaxCode(), institutionProxyInfo, categoryProxyInfo);
-                institutionToReturn = institutionConnector.save(institutionEC);
-            } catch (Exception e) {
-                throw new MsCoreException(CREATE_INSTITUTION_ERROR.getMessage(), CREATE_INSTITUTION_ERROR.getCode());
-            }
-        } else {
-            institutionToReturn = opt.get();
-        }
-
         checkIfAlreadyExistsByTaxCodeAndSubunitCode(strategyInput.getTaxCode(), strategyInput.getSubunitCode());
 
+        final InstitutionPaSubunitType subunitType = strategyInput.getSubunitType();
+        final InstitutionProxyInfo institutionProxyInfo = partyRegistryProxyConnector.getInstitutionById(strategyInput.getTaxCode());
+        final CategoryProxyInfo categoryProxyInfo = partyRegistryProxyConnector.getCategory(institutionProxyInfo.getOrigin(), institutionProxyInfo.getCategory());
+
         Institution institution;
         if (InstitutionPaSubunitType.AOO.equals(subunitType)) {
-            institution = mappingToInstitutionIPAAoo(strategyInput, institutionToReturn, institutionProxyInfo, categoryProxyInfo);
+            Institution institutionEC = getOrSaveInstitutionEc(strategyInput, institutionProxyInfo, categoryProxyInfo);
+            institution = mappingToInstitutionIPAAoo(strategyInput, institutionEC, institutionProxyInfo, categoryProxyInfo);
         } else if(InstitutionPaSubunitType.UO.equals(subunitType)) {
-            institution = mappingToInstitutionIPAUo(strategyInput, institutionToReturn, institutionProxyInfo, categoryProxyInfo);
+            Institution institutionEC = getOrSaveInstitutionEc(strategyInput, institutionProxyInfo, categoryProxyInfo);
+            institution = mappingToInstitutionIPAUo(strategyInput, institutionEC, institutionProxyInfo, categoryProxyInfo);
         }else{
             log.info("createInstitution :: unsupported subunitType {}", subunitType);
-            return institutionToReturn;
+            return getOrSaveInstitutionEc(strategyInput);
```

Comment:
you could use here getInstitutionEC, something like this:

institution = getInstitutionEC(strategyInput.getTaxCode(), institutionProxyInfo, categoryProxyInfo);

so you avoid to create another method getOrSaveInstitutionEc and executing twice the check if institution already exists